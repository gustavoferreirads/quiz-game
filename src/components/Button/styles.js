import { StyleSheet } from 'react-native';
import { colors, fonts } from '../../styles';

const styles = StyleSheet.create({
  container: {
    height: 40,
    borderRadius: 30,
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 20,
    fontWeight: 'bold',
    width: 150,
  },

  'button-inline': {
    backgroundColor: colors.primary,
  },

  'button-outline': {
    backgroundColor: '#FFF',
    borderWidth: 1,
    borderColor: colors.primary,
  },

  text: {
    fontSize: fonts.big,
  },
  'text-inline': {
    color: '#FFF',
  },
  'text-outline': {
    color: colors.primary,
  },
});

export default styles;
