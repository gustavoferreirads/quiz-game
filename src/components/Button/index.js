import React, { Fragment } from 'react';
import * as PropTypes from 'prop-types';
import { Text, TouchableOpacity } from 'react-native';
import styles from './styles';

const Button = ({ style, children, type, onPress }) => (
  <TouchableOpacity style={[styles.container, style, styles[`button-${type}`]]} onPress={onPress}>
    <Text style={[styles.text, styles[`text-${type}`]]}>{children}</Text>
  </TouchableOpacity>
);

Button.defaultProps = {
  type: 'inline',
  onPress: () => {},
  style: {},
  children: <Fragment />,
};

Button.propTypes = {
  children: PropTypes.element,
  style: PropTypes.object,
  type: PropTypes.string,
  onPress: PropTypes.func,
};

export default Button;
