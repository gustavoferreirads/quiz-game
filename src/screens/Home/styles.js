import { StyleSheet } from 'react-native';
import { colors, fonts } from '../../styles';

export default StyleSheet.create({
  container: {
    flex: 1,
  },
  title: {
    fontWeight: 'bold',
    fontSize: fonts.bigger,
    textAlign: 'center',
  },
  boxTitle: {
    justifyContent: 'center',
    flex: 1,
  },
  box: {
    flexGrow: 1,
  },

  internalText: {
    color: colors.dark,
    fontSize: fonts.large,
    textAlign: 'center',
  },
  bolder: {
    fontWeight: 'bold',
  },
  grow: {
    flexGrow: 1,
  },
});
