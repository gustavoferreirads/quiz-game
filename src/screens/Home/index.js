import React from 'react';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Text, View } from 'react-native';
import * as actions from '../../redux/actions/quiz';
import Container from '../../components/Container';
import Button from '../../components/Button';

import styles from './styles';

export const Home = ({ navigation, fetchQuestions }) => {
  const navigateBegin = () => {
    fetchQuestions();
    navigation.navigate('Quiz');
  };

  return (
    <Container style={styles.container}>
      <View style={styles.boxTitle}>
        <Text style={styles.title}>Welcome to the</Text>
        <Text style={styles.title}>Trivia Challenge!</Text>
      </View>
      <View style={styles.grow}>
        <Text style={styles.internalText}>
          You will be presented with 10 True or False questions.
        </Text>
      </View>
      <View style={styles.grow}>
        <Text style={[styles.internalText, styles.bolder]}>Can you score 100%?</Text>
      </View>
      <View style={styles.grow}>
        <Button onPress={navigateBegin}>
          <Text>BEGIN</Text>
        </Button>
      </View>
    </Container>
  );
};

Home.defaultProps = {
  navigation: {},
};

Home.propTypes = {
  navigation: PropTypes.object,
  fetchQuestions: PropTypes.func.isRequired,
};

export default connect(
  () => ({}),
  dispatch => bindActionCreators(actions, dispatch),
)(Home);
