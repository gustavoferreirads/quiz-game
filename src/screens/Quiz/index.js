import React from 'react';
import * as PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Text, View } from 'react-native';
import Container from '../../components/Container';
import * as actions from '../../redux/actions/quiz';
import { withLoading } from '../../components/Loading';
import styles from './styles';
import Answer from '../../components/Answer';

const TRUE = 'True';
const FALSE = 'False';

export const Quiz = ({ navigation, current, size, doAnswer }) => {
  const { position, wrapQuestion } = current;

  const onAnswer = answer => {
    doAnswer(answer);
    if (position === size) {
      navigation.navigate('Results');
    }
  };

  return (
    <Container style={styles.container}>
      <Text style={styles.title}>{wrapQuestion.category}</Text>
      <View style={styles.content}>
        <View style={styles.box}>
          <View style={styles.internalBox}>
            <Text style={styles.question}>{wrapQuestion.question}</Text>
          </View>
          <Text style={styles.position}>
            {position} of {size}
          </Text>
        </View>
        <View style={styles.contentAnswer}>
          <Answer option="A" text="True" onPress={() => onAnswer(TRUE)} />
          <Answer option="B" text="False" onPress={() => onAnswer(FALSE)} />
        </View>
      </View>
    </Container>
  );
};

Quiz.defaultProps = {
  current: {},
  navigation: {},
  size: 0,
};

Quiz.propTypes = {
  navigation: PropTypes.object,
  doAnswer: PropTypes.func.isRequired,
  current: PropTypes.object,
  size: PropTypes.number,
};

export default withLoading(
  connect(
    store => ({
      current: store.Quiz.current,
      size: store.Quiz.size,
    }),
    dispatch => bindActionCreators(actions, dispatch),
  )(Quiz),
);
