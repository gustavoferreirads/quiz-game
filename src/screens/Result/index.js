import React, { Fragment } from 'react';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { FlatList, Text, View } from 'react-native';
import ProgressCircle from 'react-native-progress-circle';
import { AntDesign } from '@expo/vector-icons/build/Icons';
import Button from '../../components/Button';
import Container from '../../components/Container';
import * as actions from '../../redux/actions/quiz';
import { icons, progressCircle, styles } from './styles';

const { radius, borderWidth, color, shadowColor, bgColor } = progressCircle;

export const Result = ({ navigation, fetchQuestions, answers, score, percent }) => {
  const navigateBegin = () => {
    navigation.navigate('Home');
  };

  const navigateQuiz = () => {
    fetchQuestions();
    navigation.navigate('Quiz');
  };

  return (
    <Container style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.title}>You scored</Text>

        <View style={styles.graph}>
          <ProgressCircle
            percent={percent}
            radius={radius}
            borderWidth={borderWidth}
            color={color}
            shadowColor={shadowColor}
            bgColor={bgColor}
          >
            <Text style={styles.title}>
              {score}/{answers.length}
            </Text>
          </ProgressCircle>
        </View>

        <Text style={styles.subTitle}>You got {percent}% of the answers correct</Text>
        <Text style={styles.subTitle}>Congratulations!</Text>
      </View>

      <View style={styles.content}>
        <FlatList
          data={answers}
          renderItem={({ item }) => (
            <View style={styles.item}>
              <View style={styles.headerItem}>
                <Text style={styles.description}>{item.question}</Text>
                <Text style={[styles.description, styles.answer]}>{item.correct_answer}</Text>
              </View>

              <View style={styles.result}>
                {item.isCorrect && (
                  <Fragment>
                    <Text style={styles.text}>You're right!</Text>
                    <AntDesign name={icons.like.name} size={icons.size} color={icons.like.color} />
                  </Fragment>
                )}

                {!item.isCorrect && (
                  <Fragment>
                    <Text style={styles.text}>You missed!</Text>
                    <AntDesign
                      name={icons.unlike.name}
                      size={icons.size}
                      color={icons.unlike.color}
                    />
                  </Fragment>
                )}
              </View>
            </View>
          )}
          keyExtractor={(item, index) => index.toString()}
        />
      </View>
      <View style={styles.actions}>
        <Button onPress={navigateQuiz}>
          <Text>PLAY AGAIN?</Text>
        </Button>

        <Button type="outline" onPress={navigateBegin}>
          <Text>HOME</Text>
        </Button>
      </View>
    </Container>
  );
};

Result.defaultProps = {
  answers: [],
  navigation: {},
  score: 0,
  percent: 0,
};

Result.propTypes = {
  navigation: PropTypes.object,
  answers: PropTypes.arrayOf(PropTypes.object),
  fetchQuestions: PropTypes.func.isRequired,
  score: PropTypes.number,
  percent: PropTypes.number,
};

export default connect(
  ({ Quiz }) => ({
    answers: Quiz.answers,
    score: Quiz.score,
    percent: Quiz.percent,
  }),
  dispatch => bindActionCreators(actions, dispatch),
)(Result);
