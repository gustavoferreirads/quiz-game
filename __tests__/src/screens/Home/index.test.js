import React from 'react';
import { shallow } from 'enzyme';
import { Home } from '../../../../src/screens/Home';
import Button from '../../../../src/components/Button';

it('renders a Home ', () => {
  const rendered = shallow(<Home fetchQuestions={() => {}} navigation={{}} />);
  expect(rendered).toMatchSnapshot();
});

it('simulate click on button "Begin"', () => {
  const fetchQuestions = jest.fn(() => {});
  const navigator = {
    navigate: jest.fn(x => x),
  };
  const wrapper = shallow(<Home navigation={navigator} fetchQuestions={fetchQuestions} />);
  wrapper
    .find(Button)
    .first()
    .props()
    .onPress();
  expect(navigator.navigate.mock.calls[0][0]).toBe('Quiz');
  expect(navigator.navigate.mock.calls.length).toBe(1);
  expect(fetchQuestions.mock.calls.length).toBe(1);
});
