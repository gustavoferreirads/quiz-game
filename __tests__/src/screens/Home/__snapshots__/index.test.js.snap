// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Home  1`] = `
<Container
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Component
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Component
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Welcome to the
    </Component>
    <Component
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Trivia Challenge!
    </Component>
  </Component>
  <Component
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <Component
      style={
        Object {
          "color": "#333",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      You will be presented with 10 True or False questions.
    </Component>
  </Component>
  <Component
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <Component
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 20,
            "textAlign": "center",
          },
          Object {
            "fontWeight": "bold",
          },
        ]
      }
    >
      Can you score 100%?
    </Component>
  </Component>
  <Component
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <Button
      onPress={[Function]}
      style={Object {}}
      type="inline"
    >
      <Component>
        BEGIN
      </Component>
    </Button>
  </Component>
</Container>
`;
